core/lib/zksync_core/src/vm_runner/storage.rs

```diff
@@ -0,0 +1,456 @@
+/// so that they are immediately accessible by [`VmRunnerStorage`].
+#[derive(Debug)]
+pub struct StorageSyncTask<L: VmRunnerStorageLoader> {
+    pool: ConnectionPool<Core>,
+    max_batches_to_load: u32,
+    l1_batch_params_provider: L1BatchParamsProvider,
+    chain_id: L2ChainId,
+    rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+    loader: L,
+    state: Arc<RwLock<State>>,
+    catchup_task: AsyncCatchupTask,
+}
+
+impl<L: VmRunnerStorageLoader> StorageSyncTask<L> {
+    async fn new(
+        pool: ConnectionPool<Core>,
+        max_batches_to_load: u32,
+        chain_id: L2ChainId,
+        rocksdb_path: String,
+        enum_index_migration_chunk_size: usize,
+        loader: L,
+        rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+        state: Arc<RwLock<State>>,
+    ) -> anyhow::Result<Self> {
+        let mut conn = pool.connection_tagged(L::name()).await?;
+        let l1_batch_params_provider = L1BatchParamsProvider::new(&mut conn)
+            .await
+            .context("Failed initializing L1 batch params provider")?;
+        let catchup_task = AsyncCatchupTask::new(
+            pool.clone(),
+            rocksdb_path,
+            enum_index_migration_chunk_size,
+            rocksdb_cell.clone(),
+            Some(loader.latest_processed_batch(&mut conn).await?),
+        );
+        drop(conn);
+        Ok(Self {
+            pool,
+            max_batches_to_load,
+            l1_batch_params_provider,
+            chain_id,
+            rocksdb_cell,
+            loader,
+            state,
+            catchup_task,
+        })
+    }
+
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        self.catchup_task.run(stop_receiver.clone()).await?;
+        let rocksdb = self.rocksdb_cell.get().ok_or_else(|| {
+            anyhow::anyhow!("Expected RocksDB to be initialized by `AsyncCatchupTask`")
+        })?;
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("`StorageSyncTask` was interrupted");
+                return Ok(());
+            }
+            // State guard lock also serves as a Mutex between `StorageSyncTask` and `VmRunnerStorage`
+            let mut state = self.state.write().await;
+            let mut conn = self.pool.connection_tagged(L::name()).await?;
+            let latest_processed_batch = self.loader.latest_processed_batch(&mut conn).await?;
+            let next_batch = (latest_processed_batch + 1).min(
+                conn.blocks_dal()
+                    .get_sealed_l1_batch_number()
+                    .await?
+                    .unwrap_or_default(),
+            );
+            let rocksdb_builder = RocksdbStorageBuilder::from_rocksdb(rocksdb.clone());
+            let rocksdb = rocksdb_builder
+                .synchronize(&mut conn, &stop_receiver, Some(next_batch))
```

Comment:
I think you should sync storage up to `next_batch - 1`, because executing `next_batch` requires storage at the end of `next_batch - 1`.

Reply:
Ah yes fair point